File: Source/code/frames/frameLayerPanel.pas
// ===================================================================
//
// (c) Paul Alan Freshney 2012-2022
// www.freshney.org :: paul@freshney.org :: maximumoctopus.com
//
// https://github.com/MaximumOctopus/LEDMatrixStudio
//
// Please do not modifiy this comment section

//
// ===================================================================

unit frameLayerPanel;


interface


uses
  Winapi.Windows, Winapi.Messages, System.SysUtils, System.Variants, System.Classes,
  Vcl.Graphics, Vcl.Controls, Vcl.Forms, Vcl.Dialogs, Vcl.Grids, Vcl.StdCtrls,
  Vcl.Buttons, Vcl.ExtCtrls, System.UITypes, Vcl.Menus, thematrix,

  utility, languagehandler,

  matrixconstants, formAddLayer;


type
  TLayerMenu = procedure(aItem : integer) of object;


  TframeLayers = class(TFrame)
    Panel1: TPanel;
    sbLayersRename: TSpeedButton;
    sbDeleteLayer: TSpeedButton;
    sbAddNewLayer: TSpeedButton;
    sbLayerPanelClose: TSpeedButton;
    sbLayerUp: TSpeedButton;
    sbLayerDown: TSpeedButton;
    cbSyncAllLayers: TCheckBox;
    sgLayers: TStringGrid;
    puLayerTable: TPopupMenu;
    Clearselectedlayer1: TMenuItem;
    procedure sbLayersRenameClick(Sender: TObject);
    procedure sbAddNewLayerClick(Sender: TObject);
    procedure sbDeleteLayerClick(Sender: TObject);
    procedure sbLayerUpClick(Sender: TObject);
    procedure sbLayerDownClick(Sender: TObject);
    procedure sbLayerPanelCloseClick(Sender: TObject);
    procedure sgLayersClick(Sender: TObject);
    procedure sgLayersSelectCell(Sender: TObject; ACol, ARow: Integer;
      var CanSelect: Boolean);
    procedure Clearselectedlayer1Click(Sender: TObject);
  private
    FOnClose : TNotifyEvent;
    FOnMenu  : TLayerMenu;

    procedure SetGUILanguageText;

    procedure SetSyncAll(aSync : boolean);
    function  GetSyncAll: boolean;
  public
    procedure UpdateLayerTable;

    property OnClose : TNotifyEvent read FOnClose   write FOnClose;
    property OnMenu  : TLayerMenu   read FOnMenu    write FOnMenu;
    property SyncAll : boolean      read GetSyncAll write SetSyncAll;
  end;


implementation


{$R *.dfm}


const
  CCellVisible = 0;
  CCellLocked  = 1;
  CCellName    = 2;


procedure TframeLayers.SetGUILanguageText;
begin
  cbSyncAllLayers.Caption := GLanguageHandler.Text[kSyncAll];
end;


procedure TframeLayers.UpdateLayerTable;
var
  t, lCount : integer;

begin
  lCount            := MatrixMain.GetLayerCount;

  sgLayers.RowCount := lCount + 1;

  for t := 0 to lCount - 1 do begin
    sgLayers.Cells[CCellName, 1 + lCount - 1 - t] := MatrixMain.GetLayerName(t);

    if MatrixMain.IsVisible(t) then
      sgLayers.Cells[CCellVisible, 1 + lCount - 1 - t] := 'x'
    else begin
      if lCount = 1 then begin
        MatrixMain.SetVisibility(t, True);

        sgLayers.Cells[CCellVisible, 1 + lCount - 1 - t] := 'x';
      end
      else
        sgLayers.Cells[CCellVisible, 1 + lCount - 1 - t] := '';
    end;

    if MatrixMain.IsLayerLocked(t) then
      sgLayers.Cells[CCellLocked, 1 + lCount - 1 - t] := 'L'
    else
      sgLayers.Cells[CCellLocked, 1 + lCount - 1 - t] := '';
  end;

  if (MatrixMain.SoftwareMode = smAnimation) and (lCount > 1) then
    sbDeleteLayer.Enabled := True
  else
    sbDeleteLayer.Enabled := False;

  if (MatrixMain.SoftwareMode = smAnimation) then
    sbAddNewLayer.Enabled := True
  else
    sbAddNewLayer.Enabled := False;

  sgLayers.Selection := TGridRect(Rect(0, 1 + lCount - 1 - MatrixMain.CurrentLayer, 1, 1 + lCount - 1 - MatrixMain.CurrentLayer));

  if (MatrixMain.CurrentLayer = 0) then
    sbLayerDown.Enabled := False
  else
    sbLayerDown.Enabled := True;

  if (MatrixMain.CurrentLayer = MatrixMain.GetLayerCount - 1) then
    sbLayerUp.Enabled := False
  else
    sbLayerUp.Enabled := True;
end;


procedure TframeLayers.SetSyncAll(aSync : boolean);
begin
  cbSyncAllLayers.Checked := True;
end;


procedure TframeLayers.sgLayersClick(Sender: TObject);
begin
  MatrixMain.CurrentLayer := sgLayers.RowCount - 1 - sgLayers.Selection.Top;

  if (MatrixMain.CurrentLayer = 0) then
    sbLayerDown.Enabled := False
  else
    sbLayerDown.Enabled := True;

  if (MatrixMain.CurrentLayer = MatrixMain.GetLayerCount - 1) then
    sbLayerUp.Enabled := False
  else
    sbLayerUp.Enabled := True;
end;


procedure TframeLayers.sgLayersSelectCell(Sender: TObject; ACol, ARow: Integer;
  var CanSelect: Boolean);
var
  lLayer : integer;

begin
  CanSelect := True;

  if (ACol = 0) and (ARow > 0) then begin
    lLayer := sgLayers.RowCount - 1 - ARow;

    if sgLayers.Cells[ACol, ARow] = 'x' then begin
      sgLayers.Cells[ACol, ARow] := '';

      MatrixMain.SetVisibility(lLayer, False);
    end
    else begin
      sgLayers.Cells[ACol, ARow] := 'x';

      MatrixMain.SetVisibility(lLayer, True);
    end;
  end
  else if (ACol = 1) and (ARow > 0) then begin
    lLayer := sgLayers.RowCount - 1 - ARow;

    if sgLayers.Cells[ACol, ARow] = 'L' then begin
      sgLayers.Cells[ACol, ARow] := '';

      MatrixMain.UnlockLayer(lLayer);
    end
    else begin
      sgLayers.Cells[ACol, ARow] := 'L';

      MatrixMain.LockLayer(lLayer);
    end;
  end;
end;


procedure TframeLayers.Clearselectedlayer1Click(Sender: TObject);
begin
  if Assigned(FOnMenu) then
    FOnMenu(TMenuItem(Sender).Tag);
end;


function TframeLayers.GetSyncAll: boolean;
begin
  Result := cbSyncAllLayers.Checked;
end;


procedure TframeLayers.sbAddNewLayerClick(Sender: TObject);
var
  t : integer;
  lLayers : TStringList;
  lAddObject : TAddObject;

begin
  lLayers := TStringList.Create;

  for t := 0 to MatrixMain.GetLayerCount - 1 do
    lLayers.Add(MatrixMain.GetLayerName(t));

  lAddObject := DoAddLayer(lLayers);

  if (lAddObject.Process) then begin
    if lAddObject.Name <> '' then begin
      if lAddObject.CopyFrom then
        MatrixMain.AddLayerAsCopy(lAddObject.Name, lAddObject.CopyLayer)
      else
        MatrixMain.AddLayer(lAddObject.Name);
    end;
  end;

  lLayers.Free;
end;


procedure TframeLayers.sbDeleteLayerClick(Sender: TObject);
begin
  if MessageDlg(GLanguageHandler.Text[kDeleteCurrentLayer] + #13#13 + GLanguageHandler.Text[kThisCannotBeUndone], mtWarning, mbYesNo, 0) = mrYes then begin
    MatrixMain.DeleteLayer(sgLayers.RowCount - 1 - sgLayers.Selection.Top);
  end;
end;


procedure TframeLayers.sbLayerDownClick(Sender: TObject);
begin
  MatrixMain.MoveDown(MatrixMain.CurrentLayer);
end;


procedure TframeLayers.sbLayerPanelCloseClick(Sender: TObject);
begin
 if (Assigned(FOnClose)) then
   FOnClose(Self);
end;


procedure TframeLayers.sbLayersRenameClick(Sender: TObject);
var
  lLayerName : string;

begin
  if (sgLayers.Selection.Top >= 1) then begin

    lLayerName := InputBox(GLanguageHandler.Text[kChangeLayerName], GLanguageHandler.Text[kNewName], sgLayers.Cells[CCellLocked, sgLayers.Selection.Top]);

    if lLayerName <> '' then begin
      MatrixMain.SetLayerName(sgLayers.RowCount - 1 - sgLayers.Selection.Top, lLayerName);
    end;
  end;
end;


procedure TframeLayers.sbLayerUpClick(Sender: TObject);
begin
  MatrixMain.MoveUp(MatrixMain.CurrentLayer);
end;


end.

File: Source/Frames/FrameLayerPanel.cpp
// ===================================================================
//
//   (c) Paul Alan Freshney 2012-2024
//   www.freshney.org :: paul@freshney.org :: maximumoctopus.com
//
//   https://github.com/MaximumOctopus/LEDMatrixStudio
//
//   https://maximumoctopus.hashnode.dev/
//
//   C++ Rewrite October 11th 2023
//
// ===================================================================

#include <vcl.h>
#pragma hdrstop

#include "FrameLayerPanel.h"
#include "LanguageConstants.h"
#include "LanguageHandler.h"

#include "FormAddLayer.h"
#include "Utility.h"

extern LanguageHandler *GLanguage;

#pragma package(smart_init)
#pragma resource "*.dfm"
TframeLayers *frameLayers;

__fastcall TframeLayers::TframeLayers(TComponent* Owner)
	: TFrame(Owner)
{
}


void __fastcall TframeLayers::sbLayerUpClick(TObject *Sender)
{
	ParentMatrix->MoveUp(ParentMatrix->GetCurrentLayer());
}


void __fastcall TframeLayers::sbLayersRenameClick(TObject *Sender)
{
	if (sgLayers->Selection.Top >= 1)
	{
		std::wstring name = InputBox(GLanguageHandler->Text[kChangeLayerName].c_str(),
									 GLanguageHandler->Text[kNewName].c_str(),
									 sgLayers->Cells[CCellLocked][sgLayers->Selection.Top]).c_str();

		if (!name.empty())
		{
			ParentMatrix->SetLayerName(name, sgLayers->RowCount - 1 - sgLayers->Selection.Top);
		}
	}
}


void __fastcall TframeLayers::sbLayerPanelCloseClick(TObject *Sender)
{
	if (OnClose) OnClose(this);
}


void __fastcall TframeLayers::sbLayerDownClick(TObject *Sender)
{
	ParentMatrix->MoveDown(ParentMatrix->GetCurrentLayer());
}


void __fastcall TframeLayers::sbAddNewLayerClick(TObject *Sender)
{
	std::vector<std::wstring> layers;

	for (int t = 0; t < ParentMatrix->GetLayerCount(); t++)
	{
		layers.push_back(ParentMatrix->GetLayerName(t));
	}

	AddObject ao = OpenAddLayer(layers);

	if (ao.Process)
	{
		if (!ao.Name.empty())
		{
			if (ao.CopyFrom)
			{
				ParentMatrix->AddLayerAsCopy(ao.Name, ao.CopyLayer);
			}
			else
			{
				ParentMatrix->AddLayer(ao.Name);
			}
		}
	}
}


void __fastcall TframeLayers::sbDeleteLayerClick(TObject *Sender)
{
	if (MessageDlg(Utility::WS2US(GLanguageHandler->Text[kDeleteCurrentLayer] + L"\n\n" + GLanguageHandler->Text[kThisCannotBeUndone]), mtWarning, mbYesNo, 0) == mrYes)
	{
		ParentMatrix->DeleteLayer(sgLayers->RowCount - 1 - sgLayers->Selection.Top);
	}
}


void __fastcall TframeLayers::sgLayersClick(TObject *Sender)
{
	ParentMatrix->SetCurrentLayer(sgLayers->RowCount - 1 - sgLayers->Selection.Top);

	if (ParentMatrix->GetCurrentLayer() == 0)
	{
		sbLayerDown->Enabled = false;
	}
	else
	{
		sbLayerDown->Enabled = true;
	}

	if (ParentMatrix->GetCurrentLayer() == ParentMatrix->GetLayerCount() - 1)
	{
		sbLayerUp->Enabled = false;
	}
	else
	{
		sbLayerUp->Enabled = true;
	}
}


void __fastcall TframeLayers::sgLayersSelectCell(TObject *Sender, int ACol, int ARow, bool &CanSelect)
{
	CanSelect = true;

	int layer = 0;

	if (ACol == 0 && ARow > 0)
	{
		layer = sgLayers->RowCount - 1 - ARow;

		if (sgLayers->Cells[ACol][ARow] == L"x")
		{
			sgLayers->Cells[ACol][ARow] = L"";

			ParentMatrix->SetVisibility(layer, false);
		}
		else
		{
			sgLayers->Cells[ACol][ARow] = L"x";

			ParentMatrix->SetVisibility(layer, true);
		}
	}
	else if (ACol == 1 && ARow > 0)
	{
		layer = sgLayers->RowCount - 1 - ARow;

		if (sgLayers->Cells[ACol][ARow] == L"L")
		{
			sgLayers->Cells[ACol][ARow] = L"";

			ParentMatrix->UnlockLayer(layer);
		}
		else
		{
			sgLayers->Cells[ACol][ARow] = L"L";

			ParentMatrix->LockLayer(layer);
		}
	}
}


void __fastcall TframeLayers::Clearselectedlayer1Click(TObject *Sender)
{
	if (OnMenu)
	{
		TMenuItem *mi = (TMenuItem*)Sender;

		OnMenu(mi->Tag);
	}
}


void TframeLayers::SetGuiLanguageText()
{
	cbSyncAllLayers->Caption = GLanguageHandler->Text[kSyncAll].c_str();
}


void TframeLayers::UpdateLayerTable()
{
	int count = ParentMatrix->GetLayerCount();

	sgLayers->RowCount = count + 1;

	for (int t = 0; t < count; t++)
	{
		sgLayers->Cells[CCellName][1 + count - 1 - t] = ParentMatrix->GetLayerName(t).c_str();

		if (ParentMatrix->IsVisible(t))
		{
			sgLayers->Cells[CCellVisible][1 + count - 1 - t] = L"x";
		}
		else
		{
			if (count == 1)
			{
				ParentMatrix->SetVisibility(t, true);

				sgLayers->Cells[CCellVisible][1 + count - 1 - t] = L"x";
			}
			else
			{
				sgLayers->Cells[CCellVisible][1 + count - 1 - t] = L"";
			}
		}

		if (ParentMatrix->IsLayerLocked(t))
		{
			sgLayers->Cells[CCellLocked][1 + count - 1 - t] = L"L";
		}
		else
		{
			sgLayers->Cells[CCellLocked][1 + count - 1 - t] = L"";
		}
	}

	if (ParentMatrix->GetSoftwareMode() == SoftwareMode::kAnimation && count > 1)
	{
		sbDeleteLayer->Enabled = true;
	}
	else
	{
		sbDeleteLayer->Enabled = false;
	}

	if (ParentMatrix->GetSoftwareMode() == SoftwareMode::kAnimation)
	{
		sbAddNewLayer->Enabled = true;
	}
	else
	{
		sbAddNewLayer->Enabled = false;
	}

	TGridRect selection;

	selection.Left = 0;
	selection.Top = 1 + count - 1 - ParentMatrix->GetCurrentLayer();
	selection.Right = 1;
	selection.Bottom = 1 + count - 1 - ParentMatrix->GetCurrentLayer();

	sgLayers->Selection = selection;

	if (ParentMatrix->GetCurrentLayer() == 0)
	{
		sbLayerDown->Enabled = false;
	}
	else
	{
		sbLayerDown->Enabled = true;
	}

	if (ParentMatrix->GetCurrentLayer() == ParentMatrix->GetLayerCount() - 1)
	{
		sbLayerUp->Enabled = false;
	}
	else
	{
		sbLayerUp->Enabled = true;
	}
}


// updates the lock/visible status of currently available layers
void TframeLayers::UpdateExisting()
{
	int count = ParentMatrix->GetLayerCount();

	sgLayers->RowCount = count + 1;

	for (int t = 0; t < count; t++)
	{
		sgLayers->Cells[CCellName][1 + count - 1 - t] = ParentMatrix->GetLayerName(t).c_str();

		if (ParentMatrix->IsVisible(t))
		{
			sgLayers->Cells[CCellVisible][1 + count - 1 - t] = L"x";
		}
		else
		{
			if (count == 1)
			{
				ParentMatrix->SetVisibility(t, true);

				sgLayers->Cells[CCellVisible][1 + count - 1 - t] = L"x";
			}
			else
			{
				sgLayers->Cells[CCellVisible][1 + count - 1 - t] = L"";
			}
		}

		if (ParentMatrix->IsLayerLocked(t))
		{
			sgLayers->Cells[CCellLocked][1 + count - 1 - t] = L"L";
		}
		else
		{
			sgLayers->Cells[CCellLocked][1 + count - 1 - t] = L"";
		}
	}
}


void TframeLayers::SetSyncAll(bool sync)
{
	cbSyncAllLayers->Checked = sync;
}


bool TframeLayers::GetSyncAll()
{
	return cbSyncAllLayers->Checked;
}
